pkg/dns/dns_test.go

```diff
@@ -0,0 +1,377 @@
+			endpoints[endpointAddress.IP] = true
+		}
+	}
+	assert.Equal(t, len(endpoints), len(records))
+	for _, record := range records {
+		_, found := endpoints[record.Host]
+		assert.True(t, found)
+	}
+}
+
+func assertRecordPortsMatchPort(t *testing.T, port int32, records []skymsg.Service) {
+	for _, record := range records {
+		assert.Equal(t, port, int32(record.Port))
+	}
+}
+
+func assertCNameRecordsMatchEndpointIPs(t *testing.T, kd *KubeDNS, e []kapi.EndpointAddress, records []skymsg.Service) {
+	endpoints := map[string]bool{}
+	for _, endpointAddress := range e {
+		endpoints[endpointAddress.IP] = true
+	}
+	assert.Equal(t, len(e), len(records), "unexpected record count")
+	for _, record := range records {
+		_, found := endpoints[getIPForCName(t, kd, record.Host)]
+		assert.True(t, found, "Did not endpoint with address:%s", record.Host)
+	}
+}
+
+func getIPForCName(t *testing.T, kd *KubeDNS, cname string) string {
+	records, err := kd.Records(cname, false)
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(records), "Could not get IP for CNAME record for %s", cname)
+	assert.NotNil(t, net.ParseIP(records[0].Host), "Invalid IP address %q", records[0].Host)
+	return records[0].Host
+}
+
+func assertNoDNSForHeadlessService(t *testing.T, kd *KubeDNS, s *kapi.Service) {
+	records, err := kd.Records(getServiceFQDN(kd, s), false)
+	require.Error(t, err)
+	assert.Equal(t, 0, len(records))
+}
+
+func assertSRVForNamedPort(t *testing.T, kd *KubeDNS, s *kapi.Service, portName string) {
+	records, err := kd.Records(getSRVFQDN(kd, s, portName), false)
+	require.NoError(t, err)
+	assert.Equal(t, 1, len(records))
+	assert.Equal(t, getServiceFQDN(kd, s), records[0].Host)
+}
+
+func assertNoSRVForNamedPort(t *testing.T, kd *KubeDNS, s *kapi.Service, portName string) {
+	records, err := kd.Records(getSRVFQDN(kd, s, portName), false)
+	require.Error(t, err)
+	assert.Equal(t, 0, len(records))
+}
+
+func assertNoDNSForClusterIP(t *testing.T, kd *KubeDNS, s *kapi.Service) {
+	records, err := kd.Records(getServiceFQDN(kd, s), false)
+	require.Error(t, err)
+	assert.Equal(t, 0, len(records))
+}
+
+func assertDNSForClusterIP(t *testing.T, kd *KubeDNS, s *kapi.Service) {
+	serviceFQDN := getServiceFQDN(kd, s)
+	queries := []string{
+		serviceFQDN,
+		strings.Replace(serviceFQDN, ".svc.", ".*.", 1),
+		strings.Replace(serviceFQDN, s.Namespace, "*", 1),
+		strings.Replace(strings.Replace(serviceFQDN, s.Namespace, "*", 1), ".svc.", ".*.", 1),
+		"*." + serviceFQDN,
+	}
+	for _, query := range queries {
+		records, err := kd.Records(query, false)
+		require.NoError(t, err)
+		assert.Equal(t, 1, len(records))
+		assert.Equal(t, s.Spec.ClusterIP, records[0].Host)
+	}
+}
+
+func getServiceFQDN(kd *KubeDNS, s *kapi.Service) string {
+	return fmt.Sprintf("%s.%s.svc.%s", s.Name, s.Namespace, kd.domain)
+}
+
+func getEndpointsFQDN(kd *KubeDNS, e *kapi.Endpoints) string {
+	return fmt.Sprintf("%s.%s.svc.%s", e.ObjectMeta.Name, e.ObjectMeta.Namespace, kd.domain)
```

Comment:
Can't the ".ObjectMeta" part be dropped just like in the previous function? It is arguably easier to read.